### ISSUE TITLE
All props for structs should be enforced to be inputted by exposing functions not structs

### ISSUE THREAD

**Issue by youssefsharief** [Synthesis-AI-Dev/scenereduce#15]
Golang does not ensure that structs are instantiated with all props. For example, a float parameter that is not initialized is set to be 0 under the hood which causes problems like this [one](https://github.com/Synthesis-AI-Dev/scenereduce/issues/14) and increase complexity in scenereduce since we need to check whether a prop is inputted or not. I believe it is better to use factory methods, instead. The first section of this [article](https://www.sohamkamani.com/golang/2018-06-20-golang-factory-patterns/) explains what I mean.

**Comment by rkotcher-synthesisai** [comment]
Yes, agreed.

Could you provide a specific case where this was causing issues, though?

In some cases, I'm using pointers to ensure that zero-value is not treated as a real value. For example, by making Int.Max a pointer, we don't end up thinking that max is 0 just because it isn't set.

To your point though, we do use factory pattern in apis, and I believe we should do that here too. I wonder if we should move scenereduce into apis first? that might make the migration a bit cleaner since we are both working on scenereduce right now. what do you think?

**Comment by youssefsharief** [comment]
[This is one case](https://github.com/Synthesis-AI-Dev/scenereduce/issues/14) were not enforcing inputting all props was causing problems. 

I first tried to set max to a pointer, but then I found that we would need to change how people API would call the struct. For example, instead of inputting `100.0`, we would need to input a pointer to it. Then, I thought it would be better to use the factory pattern and forget about all these problems (even for different cases).

I think we should move scenereduce to apis first then migrate to the factory pattern. I think it will be easier to migrate to the factory pattern while `scenereduce` is in `apis` repo.
